### PR TITLE
fix #4675: adding an informer timeout handled directly in our client

### LINKED ISSUE #4675: Set a timeout client side in informer watches
> I am wondering why Fabric8 is not specifying default timeout for watch requests?

Its never been requested as a feature.  The understanding is that api server will enforce its default timeout (30 minutes I believe), with some jitter of its own.

_Originally posted by @shawkins in https://github.com/fabric8io/kubernetes-client/issues/4670#issuecomment-1348622780_

### PR DESCRIPTION
**Pull request by shawkins**
## Description

Fix #4675

This seems like a little less work than updating the mock and other tests with fixed expectations.  Here we'll trigger the shutdown of the websocket after 5 - 10 minutes (the same as the go client).

Another thought is to eliminate the random strategy specifically for the mock client, but that also seems hackish.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
In this PR we're killing the Watch client-wise as opposed to sending a parameter with a timeout. This is because our MockServer doesn't support parameters with variable values.

We'll probably want to follow up **in the future** with the improvement of the MockServer to allow for variable path parameters and then sending the timeout parameter in the URL for the Watch request including the jitter.

**Comment by shawkins** [comment]
Addressed the conflict and also added a comment / refinement about the message processing behavior - what matters is how things are handled at the watchmanager.  We do not, and should not, try and guard against further message processing in the reflector.

**Comment by shawkins** [comment]
After reviewing again decided this is best handled by an existing method on the http / websocket watch handlers - closeRequest is already supposed to trigger a graceful restart.  Doing this via the exception mechanism is not as good as it forces a re-list as well.

**Comment by shawkins** [comment]
One, hopefully last, refinement which also is now setting the url timeout as well as a fallback.  This also highlights the scope of the hard-coded expectation issue.  There are close to 40 places where this appears and several of those are shared among multiple tests.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManager.java [hunk 1]
```diff
@@ -87,9 +97,20 @@
...
+  protected abstract void start(URL url, Map<String, String> headers, WatchRequestState state);
+
+  /**
+   * Attempt to gracefully close the current request.
+   * <p>
+   * If forceClosed has not been set, then it's expected that the watch will
+   * attempt to reconnect
+   */
+  public void closeRequest() {
+    Optional.ofNullable(latestRequestState).ifPresent(s -> s.closed.set(true));
```
Comment (sunix) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManager.java:109]:
```suggestion
    Optional.ofNullable(latestRequestState).ifPresent(state -> state.closed.set(true));
```

**Review by sunix — APPROVED** [review]

**Review by manusa — APPROVED** [review]

**Pull request merged**